test/integration/dualstack/dualstack_endpoints_test.go

```diff
@@ -0,0 +1,294 @@
+					Ports: []v1.ServicePort{
+						{
+							Name:       fmt.Sprintf("port-test-%d", i),
+							Port:       443,
+							TargetPort: intstr.IntOrString{IntVal: 443},
+							Protocol:   "TCP",
+						},
+					},
+				},
+			}
+
+			// create a service
+			_, err = client.CoreV1().Services(ns.Name).Create(context.TODO(), svc, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("error creating service: %v", err)
+			}
+
+			// wait until endpoints are created
+			// legacy endpoints are not dual stack
+			// and use the address of the first IP family
+			if err := wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				e, err := client.CoreV1().Endpoints(ns.Name).Get(context.TODO(), svc.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Logf("error fetching endpoints: %v", err)
+					return false, nil
+				}
+				// check if the endpoint addresses match the pod IP of the first IPFamily of the service
+				// since this is an integration test PodIPs are not "ready"
+				if len(e.Subsets) > 0 && len(e.Subsets[0].NotReadyAddresses) > 0 {
+					if e.Subsets[0].NotReadyAddresses[0].IP == podIPbyFamily[tc.ipFamilies[0]] {
+						return true, nil
+					}
+					t.Logf("endpoint address %s does not match PodIP %s ", e.Subsets[0].Addresses[0].IP, podIPbyFamily[tc.ipFamilies[0]])
+				}
+				t.Logf("endpoint does not contain addresses: %s", e.Name)
+				return false, nil
+			}); err != nil {
+				t.Fatalf("endpoints not found: %v", err)
+			}
+
+			// wait until the endpoint slices are created
+			err = wait.PollImmediate(1*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				lSelector := discovery.LabelServiceName + "=" + svc.Name
+				esList, err := client.DiscoveryV1beta1().EndpointSlices(ns.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				if err != nil {
+					t.Logf("error listing EndpointSlices: %v", err)
+					return false, err
+				}
+				// there must be an endpoint slice per ipFamily
+				if len(esList.Items) != len(tc.ipFamilies) {
+					t.Logf("waiting for EndpointSlice to be created %v", esList)
+					return false, nil
+				}
+				// there must be an endpoint address per each IP family
+				for _, ipFamily := range tc.ipFamilies {
+					found := false
+					for _, slice := range esList.Items {
+						// check if the endpoint addresses match the pod IPs
+						if len(slice.Endpoints) > 0 && len(slice.Endpoints[0].Addresses) > 0 {
+							if string(ipFamily) == string(slice.AddressType) &&
+								slice.Endpoints[0].Addresses[0] == podIPbyFamily[ipFamily] {
+								found = true
+								break
+							}
+						}
+						t.Logf("waiting endpoint slice to contain addresses")
+					}
+					if !found {
+						t.Logf("endpoint slices does not contain PodIP %s", podIPbyFamily[ipFamily])
+						return false, nil
+					}
+				}
+				return true, nil
+			})
+			if err != nil {
+				t.Fatalf("timed out waiting for endpoint slices: %v", err)
```

Comment:
i think err at this point would already say "timed out waiting for condition..."

this could be:
> error waiting for slices: %v